### PR TITLE
arbitrary sidecar support

### PR DESCRIPTION
**Pull request by theRealWardo**
see the discussion in #264

### CODE REVIEW

File: pkg/cluster/k8sres.go [hunk 1]
```diff
@@ -525,6 +527,52 @@ func (c *Cluster) generatePodTemplate(
...
+							FieldRef: &v1.ObjectFieldSelector{
+								APIVersion: "v1",
+								FieldPath:  "metadata.namespace",
+							},
+						},
+					},
+					// TODO: figure out auth.
+					/*
+						{
+							Name:  "POSTGRES_USER",
```
Comment (alexeyklyukin) [inline pkg/cluster/k8sres.go]: You can expose them from the secrets, i.e.
```
ValueFrom: &v1.EnvVarSource{
				SecretKeyRef: &v1.SecretKeySelector{
					LocalObjectReference: v1.LocalObjectReference{
						Name: c.credentialSecretName(c.OpConfig.SuperUsername),
					},
					Key: "password",
				},
			},
```

File: pkg/cluster/k8sres.go [hunk 2]
```diff
@@ -525,6 +527,52 @@ func (c *Cluster) generatePodTemplate(
...
+	monitoringContainerImage := c.OpConfig.MonitoringDockerImage
+	if monitoringDockerImage != nil && *monitoringDockerImage != "" {
+		monitoringContainerImage = *monitoringDockerImage
+	}
+	if monitoringContainerImage != "" {
+		podSpec.Containers = append(
+			podSpec.Containers,
+			v1.Container{
+				Name:            "monitoring-sidecar",
+				Image:           c.monitoringImage(),
```
Comment (alexeyklyukin) [inline pkg/cluster/k8sres.go]: Looks like it is not defined anywhere.

File: pkg/cluster/k8sres.go (hunk 1 above)
Comment (theRealWardo) [reply pkg/cluster/k8sres.go]: done!

File: pkg/cluster/k8sres.go (hunk 2 above)
Comment (theRealWardo) [reply pkg/cluster/k8sres.go]: yea this was a sketch and I was trying to figure out how to get my environment set up... all fixed up and working now.

**Comment by theRealWardo** [comment]
okay so I tried to do something like `Sidecars []v1.Container` rather than a custom `Sidecar` type, but there were some problems specifically with the handling of resources there.

that of course means that there's a bit more of a painful path to supporting things like adding custom volumes/mounts here, but I think this is probably a decent start. just let me know what you think!

**Comment by alexeyklyukin** [comment]
@theRealWardo thanks for updating the PR. I see little use in modifying individual cluster manifests. What if you run 100 clusters managed by the operator, would you ask each one of them to change its definition to incorporate the same container image (be it monitoring or log support)? Instead, we can pass this information via the operator configuration to configure all clusters managed by the operator at once and uniformly. It could be a parameter of the operator itself, i.e:
```yaml
sidecar_containers: "name:image, name:image, name:image". 
```
We could pass the necessary minimum of environment variables like credentials to those containers via the `c.generatePodTemplate` and alike and use `pod_environment_configmap ` to propagate the rest.

**Comment by theRealWardo** [comment]
@alexeyklyukin while I do agree there is a use case for cluster wide sidecars, I would propose supporting both. I think that there is a use case for configuring an individual cluster's sidecars differently from other clusters' sidecars.

I also believe this project has poor support for structured data in the operator config. how am I supposed to specify ports for example? the structured string in the operator config is extremely hard to read and therefore error prone. it further requires maintaining separate object definition from the standard Kubernetes objects, making the learning curve higher, and thus reducing potential traction of this operator. for these reasons, I avoided the operator config for now and have proposed a change to the cluster definition.

**Comment by theRealWardo** [comment]
@alexeyklyukin friendly ping. would love to get this or something like it in. we've been running this code in our staging environment fine for the last week or so.  😄

**Comment by alexeyklyukin** [comment]
> @alexeyklyukin while I do agree there is a use case for cluster wide sidecars, I would propose supporting both. I think that there is a use case for configuring an individual cluster's sidecars differently from other clusters' sidecars.

I see none. What is your use case for that?

> I also believe this project has poor support for structured data in the operator config. how am I supposed to specify ports for example? the structured string in the operator config is extremely hard to read and therefore error prone. it further requires maintaining separate object definition from the standard Kubernetes objects, making the learning curve higher, and thus reducing potential traction of this operator. for these reasons, I avoided the operator config for now and have proposed a change to the cluster definition.


Fair enough, I have some plans to address that.

**Comment by alexeyklyukin** [comment]
@theRealWardo I have created https://github.com/zalando-incubator/postgres-operator/pull/331 based on your PR that adds support for global containers and does some refactoring. I have considered doing it against your own branch, but since the latest master is not merged in there the PR shows too many unrelated changes.

I'd appreciate if you take a look.

**Pull request closed without merging**